File: src/206.reverse-linked-list/solution_recursive.ts
/*
 * @lc app=leetcode.cn id=206 lang=typescript
 *
 * [206] 反转链表
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

// @lc code=start
function reverseListRecursive(node: ListNode | null): ListNode | null {
  if (!node || !node.next) {
    return node;
  }

  const head = reverseListRecursive(node.next);

  node.next.next = node;
  node.next = null;

  return head;
}

function reverseList(head: ListNode | null): ListNode | null {
  return reverseListRecursive(head);
}
// @lc code=end

export default reverseList;

File: src/206.reverse-linked-list/solution_recursive.cpp
/*
 * @lc app=leetcode.cn id=206 lang=cpp
 *
 * [206] 反转链表
 */

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

// @lc code=start
class Solution {
 private:
  ListNode *reverseListRecursive(ListNode *node) {
    if (!node || !node->next) {
      return node;
    }

    ListNode *head = reverseListRecursive(node->next);

    node->next->next = node;
    node->next = nullptr;

    return head;
  }

 public:
  ListNode *reverseList(ListNode *head) { return reverseListRecursive(head); }
};
// @lc code=end
